test/src/test/integration/link_git_protocol.rs

```diff
@@ -5,19 +5,109 @@
 
 use std::{
     collections::BTreeSet,
+    convert::TryInto,
     io,
     path::Path,
     sync::{atomic::AtomicBool, Arc},
 };
 
 use bstr::ByteSlice as _;
 use futures::{AsyncReadExt as _, TryFutureExt as _};
+use git_repository as git;
+use git::{
+    prelude::*,
+    refs::{
+        transaction::{Change, Create, RefEdit},
+        Target,
+    },
+};
 use link_git_protocol::{fetch, ls, packwriter, upload_pack, ObjectId, PackWriter, Ref};
 use tempfile::{tempdir, TempDir};
 
 fn upstream() -> TempDir {
     let tmp = tempdir().unwrap();
 
+    let repo = git::init_bare(&tmp).unwrap().into_easy();
+    let auth = git::actor::Signature::now_local_or_utc("apollo", "apollo@cree.de");
+
+    let empty_tree_id = repo.write_object(&git::objs::Tree::empty().into()).unwrap();
+    let (base_id, first_commit) = {
+        let commit_object: git::objs::Object = git::objs::Commit {
+            message: "initial".into(),
+            tree: empty_tree_id.into(),
+            author: auth.clone(),
+            committer: auth.clone(),
+            encoding: None,
+            parents: Default::default(),
+            extra_headers: Default::default(),
+        }
+        .into();
+        let commit_id = repo.write_object(&commit_object).unwrap().detach();
+        repo.edit_reference(
+            RefEdit {
+                change: Change::Update {
+                    log: Default::default(),
+                    mode: Create::Only,
+                    new: Target::Peeled(commit_id),
+                },
+                name: "refs/namespaces/foo/refs/heads/main".try_into().unwrap(),
+                deref: false,
+            },
+            git::lock::acquire::Fail::Immediately,
+            Some(&auth),
+        )
+        .unwrap();
```

Comment:
There will be an equivalent `commit()`-like method soon to reduce this boilerplate.